crates/bevy_ecs/src/query/world_query.rs

```diff
@@ -40,12 +40,16 @@ use variadics_please::all_tuples;
 /// [`QueryFilter`]: crate::query::QueryFilter
 pub unsafe trait WorldQuery {
     /// The item returned by this [`WorldQuery`]
-    /// For `QueryData` this will be the item returned by the query.
+    /// For `QueryData` this will be the data retrieved by the query,
+    /// and is visible to the end user when calling e.g. `Query<Self>::get`.
+    ///
     /// For `QueryFilter` this will be either `()`, or a `bool` indicating whether the entity should be included
     /// or a tuple of such things.
+    /// Archetypal query filters (like `With`) set this to `()`, as the filtering is done by selecting the archetypes to iterate over,
```

Comment:
It's kind of weird that `QueryFilter`s even have an `Item`.  I think it's mostly because they used to be a single trait, and the change to separate them tried to minimize changes.  

Would be more clear if we moved `Item` and `fetch` from `WorldQuery` to `QueryData`, and then did all the filtering in `filter_fetch`?  (Not in this PR, of course.)  It doesn't look like `fetch` is ever called on a `QueryFilter` outside its own implementation, so that might work.

Reply:
Spun out into #17662. Fully agree.